File: golang/03-internet-cafe.go
package main

import (
	"log"
	"math/rand"
	"sync"
	"time"
)

func main() {
	rand.Seed(1)

	delay := func() int {
		duration := 5 + rand.Intn(10)
		time.Sleep(time.Duration(duration) * time.Second)
		return duration
	}

	genTourists := func(num int) <-chan interface{} {
		touristStream := make(chan interface{}, num)

		go func() {
			for i := 0; i < num; i++ {
				touristStream <- i + 1
			}
			close(touristStream)
		}()

		return touristStream
	}

	clearQueue := func(
		queueStream <-chan interface{},
		wg *sync.WaitGroup,
		i interface{},
	) {
		log.Printf("Tourist %d is online.\n", i)
		duration := delay()
		<-queueStream
		wg.Done()
		log.Printf("Tourist %d is done, having spent %d seconds online.\n", i, duration)
	}

	queue := func(
		done chan interface{},
		queueStream chan interface{},
		wg *sync.WaitGroup,
		i interface{},
	) {
		select {
		case <-done:
			return
		case queueStream <- i:
			clearQueue(queueStream, wg, i)
		}
	}

	openCafe := func(
		done chan interface{},
		touristStream <-chan interface{},
		wg *sync.WaitGroup,
		maxComputers int,
	) {
		queueStream := make(chan interface{}, maxComputers)

		for i := range touristStream {
			go func(i interface{}) {
				select {
				case <-done:
					return
				case queueStream <- i:
					clearQueue(queueStream, wg, i)
				default:
					log.Printf("Tourist %d is waiting for turn.\n", i)
					queue(done, queueStream, wg, i)
				}
			}(i)
		}
	}

	// Initialization
	numTourists := 25
	maxComputers := 8

	done := make(chan interface{})
	defer close(done)

	var wg sync.WaitGroup
	wg.Add(numTourists)
	openCafe(done, genTourists(numTourists), &wg, maxComputers)
	wg.Wait()

	log.Println("The place is empty, let's close up and go to the beach!")
}

File: javascript/03-internet-cafe.js
async function main () {
  const tourists = Array(25)
    .fill(0)
    .map((_, i) => [i + 1, Math.random()])
    .sort(([_, a], [__, b]) => a > b)
    .map(([a, _]) => a)

  const maxConcurrency = 8
  await concurrentOnline(tourists, goOnline, maxConcurrency)
  console.log(`The place is empty, let's close up and go to the beach!`)
}

async function goOnline (i) {
  console.log(`Tourist ${i} is online.`)
  const duration = await delay(random(1000, 2000))
  console.log(`Tourist ${i} is done, having spent ${duration} seconds online.`)
  return i
}

function random (min, max) {
  const delta = max - min
  return min + Math.floor(Math.random() * delta)
}

async function delay (duration) {
  return new Promise((resolve, reject) => {
    setTimeout(_ => resolve(duration), duration)
  })
}

main().then().catch(console.error)

async function concurrentOnline (arr, fn, maxConcurrency = 10, concurrency = 0, iter = 0) {
  if (!arr.length) return
  let pending = []

  arr.map(async (item) => {
    const isLimit = concurrency >= maxConcurrency
    if (isLimit) {
      pending.push(item)
      return
    }
    concurrency += 1
    await fn(item)
    concurrency -= 1
  })

  if (iter === 0) {
    pending.forEach(i => console.log(`Tourist ${i} is waiting for turn.`))
  }

  while (concurrency === maxConcurrency) {
    await delay(random(150, 300))
  }
  return concurrentOnline(pending, fn, maxConcurrency, concurrency, iter + 1)
}
